File: server/agent/src/agents/diversity.js
function enforceDiversity(rankedPlaces, options = {}) {
    const {
        maxPerCategory = 4,
        cities = [],
        minPerCity = 4,
        totalNeeded = 20,
        surfaceHiddenGems = true
    } = options;

    const categoryCounts = {};
    const cityCounts = {};
    const selected = [];
    const selectedIds = new Set();

    const CITY_CATEGORY_MINIMUMS = {
        'agra': { 'monument': 2 }
    };

    if (cities.length > 0) {
        for (const city of cities) {
            const cityLower = city.toLowerCase();
            const cityPlaces = rankedPlaces.filter(p =>
                (p.city || '').toLowerCase() === cityLower && !selectedIds.has(p.id)
            );

            const minimums = CITY_CATEGORY_MINIMUMS[cityLower] || {};
            const minimumCounts = {};
            for (const place of cityPlaces) {
                const cat = normCategory(place.category);
                if (minimums[cat] && (minimumCounts[cat] || 0) < minimums[cat] && !selectedIds.has(place.id)) {
                    if (['restaurant', 'food stall', 'food', 'sweet shop', 'cafe'].includes(cat)) continue;
                    selected.push(place);
                    selectedIds.add(place.id);
                    categoryCounts[cat] = (categoryCounts[cat] || 0) + 1;
                    cityCounts[cityLower] = (cityCounts[cityLower] || 0) + 1;
                    minimumCounts[cat] = (minimumCounts[cat] || 0) + 1;
                }
            }

            let added = Object.values(minimumCounts).reduce((s, v) => s + v, 0);
            for (const place of cityPlaces) {
                if (added >= minPerCity) break;
                if (selectedIds.has(place.id)) continue;
                const cat = normCategory(place.category);
                if (['restaurant', 'food stall', 'food', 'sweet shop', 'cafe'].includes(cat)) continue;

                selected.push(place);
                selectedIds.add(place.id);
                categoryCounts[cat] = (categoryCounts[cat] || 0) + 1;
                cityCounts[cityLower] = (cityCounts[cityLower] || 0) + 1;
                added++;
            }
        }
    }

    for (const place of rankedPlaces) {
        if (selected.length >= totalNeeded) break;
        if (selectedIds.has(place.id)) continue;

        const cat = normCategory(place.category);
        const cityLower = (place.city || '').toLowerCase();

        if ((categoryCounts[cat] || 0) >= maxPerCategory) continue;

        selected.push(place);
        selectedIds.add(place.id);
        categoryCounts[cat] = (categoryCounts[cat] || 0) + 1;
        cityCounts[cityLower] = (cityCounts[cityLower] || 0) + 1;
    }

    if (surfaceHiddenGems && selected.length < totalNeeded) {
        const hiddenGems = rankedPlaces.filter(p =>
            !selectedIds.has(p.id) &&
            (p.popularity_rank || 999) > 15 &&
            !p.highlight
        );

        const gemSlots = Math.max(1, Math.floor(totalNeeded * 0.2));
        let gemsAdded = 0;
        for (const gem of hiddenGems) {
            if (gemsAdded >= gemSlots || selected.length >= totalNeeded) break;
            const cat = normCategory(gem.category);
            if ((categoryCounts[cat] || 0) < maxPerCategory + 1) {
                selected.push(gem);
                selectedIds.add(gem.id);
                categoryCounts[cat] = (categoryCounts[cat] || 0) + 1;
                gemsAdded++;
            }
        }
    }

    return selected;
}

function getCategoryStats(places) {
    const stats = {};
    for (const p of places) {
        const cat = normCategory(p.category);
        stats[cat] = (stats[cat] || 0) + 1;
    }
    return stats;
}

function getCityStats(places) {
    const stats = {};
    for (const p of places) {
        const city = p.city || 'Unknown';
        stats[city] = (stats[city] || 0) + 1;
    }
    return stats;
}

function normCategory(cat) {
    return (cat || 'Other').toLowerCase().trim();
}

module.exports = { enforceDiversity, getCategoryStats, getCityStats };

File: server_py/agent/agents/diversity.py
"""
Diversity Enforcement — Ensures balanced place selection.
"""


def _norm_category(cat: str) -> str:
    return (cat or "Other").lower().strip()


def enforce_diversity(
    ranked_places: list[dict],
    options: dict = None,
) -> list[dict]:
    """Enforce diversity in place selection — balanced categories, cities, hidden gems."""
    options = options or {}
    max_per_category = options.get("maxPerCategory", 4)
    cities = options.get("cities", [])
    min_per_city = options.get("minPerCity", 4)
    total_needed = options.get("totalNeeded", 20)
    surface_hidden_gems = options.get("surfaceHiddenGems", True)

    category_counts: dict[str, int] = {}
    city_counts: dict[str, int] = {}
    selected: list[dict] = []
    selected_ids: set[str] = set()

    CITY_CATEGORY_MINIMUMS = {
        "agra": {"monument": 2},
    }

    FOOD_CATEGORIES = {"restaurant", "food stall", "food", "sweet shop", "cafe"}

    # Phase 1: Ensure minimum per city
    if cities:
        for city in cities:
            city_lower = city.lower()
            city_places = [
                p for p in ranked_places
                if (p.get("city") or "").lower() == city_lower and p.get("id") not in selected_ids
            ]

            minimums = CITY_CATEGORY_MINIMUMS.get(city_lower, {})
            minimum_counts: dict[str, int] = {}

            for place in city_places:
                cat = _norm_category(place.get("category"))
                if (
                    cat in minimums
                    and minimum_counts.get(cat, 0) < minimums[cat]
                    and place.get("id") not in selected_ids
                ):
                    if cat in FOOD_CATEGORIES:
                        continue
                    selected.append(place)
                    selected_ids.add(place["id"])
                    category_counts[cat] = category_counts.get(cat, 0) + 1
                    city_counts[city_lower] = city_counts.get(city_lower, 0) + 1
                    minimum_counts[cat] = minimum_counts.get(cat, 0) + 1

            added = sum(minimum_counts.values())
            for place in city_places:
                if added >= min_per_city:
                    break
                if place.get("id") in selected_ids:
                    continue
                cat = _norm_category(place.get("category"))
                if cat in FOOD_CATEGORIES:
                    continue

                selected.append(place)
                selected_ids.add(place["id"])
                category_counts[cat] = category_counts.get(cat, 0) + 1
                city_counts[city_lower] = city_counts.get(city_lower, 0) + 1
                added += 1

    # Phase 2: Fill remaining slots with diversity constraint
    for place in ranked_places:
        if len(selected) >= total_needed:
            break
        if place.get("id") in selected_ids:
            continue

        cat = _norm_category(place.get("category"))
        if category_counts.get(cat, 0) >= max_per_category:
            continue

        selected.append(place)
        selected_ids.add(place["id"])
        category_counts[cat] = category_counts.get(cat, 0) + 1
        city_lower = (place.get("city") or "").lower()
        city_counts[city_lower] = city_counts.get(city_lower, 0) + 1

    # Phase 3: Surface hidden gems
    if surface_hidden_gems and len(selected) < total_needed:
        hidden_gems = [
            p for p in ranked_places
            if p.get("id") not in selected_ids
            and (p.get("popularity_rank") or 999) > 15
            and not p.get("highlight")
        ]

        gem_slots = max(1, int(total_needed * 0.2))
        gems_added = 0
        for gem in hidden_gems:
            if gems_added >= gem_slots or len(selected) >= total_needed:
                break
            cat = _norm_category(gem.get("category"))
            if category_counts.get(cat, 0) < max_per_category + 1:
                selected.append(gem)
                selected_ids.add(gem["id"])
                category_counts[cat] = category_counts.get(cat, 0) + 1
                gems_added += 1

    return selected


def get_category_stats(places: list[dict]) -> dict[str, int]:
    stats: dict[str, int] = {}
    for p in places:
        cat = _norm_category(p.get("category"))
        stats[cat] = stats.get(cat, 0) + 1
    return stats


def get_city_stats(places: list[dict]) -> dict[str, int]:
    stats: dict[str, int] = {}
    for p in places:
        city = p.get("city") or "Unknown"
        stats[city] = stats.get(city, 0) + 1
    return stats
